sensor/kubernetes/listener/resources/selector.go

```diff
@@ -1,45 +1,153 @@
+	return l.labels[label]
+}
+
+func (l labelWithLenImpl) Len() uint {
+	return uint(len(l.labels))
+}
+
+func createLabelsWithLen(labels map[string]string) labelWithLenImpl {
+	return labelWithLenImpl{labels}
+}
+
+// SelectorWrap holds a selector and information allowing for additional checks before matching
+type selectorWrap struct {
+	selector  labels.Selector
+	numLabels uint
+	matchNil  bool
+}
+
+func (s selectorWrap) Matches(labels labelsWithLen) bool {
+	if s.numLabels > labels.Len() {
+		return false
+	}
+	if s.numLabels == 0 {
+		return s.matchNil
+	}
+	return s.selector.Matches(labels)
+}
+
+func (s selectorWrap) getSelector() labels.Selector {
+	return s.selector
+}
+
+func (s selectorWrap) getNumLabels() uint {
+	return s.numLabels
+}
+
+func (s selectorWrap) getMatchNil() bool {
+	return s.matchNil
 }
 
 // selectorDisjunction is the disjunction (logical or) of a list of selectors.
-type selectorDisjunction []selector
+type selectorDisjunction []labels.Selector
 
-func (d selectorDisjunction) Matches(labels2 labels.Labels) bool {
+func (d selectorDisjunction) Empty() bool {
 	for _, sel := range d {
-		if sel.Matches(labels2) {
+		if !sel.Empty() {
+			return false
+		}
+	}
+	return true
+}
+
+func (d selectorDisjunction) String() string {
+	panic("unused function")
+}
+
+func (d selectorDisjunction) Add(r ...labels.Requirement) labels.Selector {
+	panic("unused function")
+}
+
+func (d selectorDisjunction) Requirements() (requirements labels.Requirements, selectable bool) {
+	panic("unused function")
+}
+
+func (d selectorDisjunction) DeepCopySelector() labels.Selector {
+	panic("unused function")
+}
+
+func (d selectorDisjunction) RequiresExactMatch(label string) (value string, found bool) {
+	panic("unused function")
+}
+
+func (d selectorDisjunction) Matches(labels labels.Labels) bool {
+	for _, sel := range d {
+		if sel.Matches(labels) {
 			return true
 		}
 	}
 	return false
 }
 
-// or returns the logical or of the given selectors.
+// or returns the logical or of the given SelectorWrappers.
 func or(sels ...selector) selector {
-	return selectorDisjunction(sels)
-}
-
-// SelectorFromMap converts the given map to a selector. It correctly translates a `nil` map to a `nothing` collector,
-// as is, e.g., used for headless services.
-func SelectorFromMap(labelsMap map[string]string) labels.Selector {
-	if len(labelsMap) == 0 {
-		return labels.Nothing()
+	var selWrapper = selectorWrap{nil, math.MaxUint, false}
+	var selectors selectorDisjunction
+	for _, s := range sels {
+		if s.getMatchNil() {
+			selWrapper.matchNil = true
+		}
+		if selWrapper.numLabels > s.getNumLabels() && (s.getNumLabels() > 0 || s.getMatchNil()) {
+			selWrapper.numLabels = s.getNumLabels()
+		}
+		selectors = append(selectors, s.getSelector())
+	}
+	if selWrapper.numLabels == math.MaxUint {
+		selWrapper.numLabels = 0
 	}
-	return labels.SelectorFromSet(labels.Set(labelsMap))
+	selWrapper.selector = selectors
+	return selWrapper
```

Comment:
What I don't like here is that the internal fields of `selectorWrap` are being computed here instead of in:
- `func (s selectorWrap) getNumLabels()` and 
- `func (s selectorWrap) getMatchNil()` respectively.

I am not sure whether it is possible to move this implementation into the respective functions, but I think we should try that.
In case this would not be possible, we should rather skip the optimization (returning earlier if the selector has more labels) and keep the code cleaner.

Reply:
The way the data structure works currently the logic has to be here, but if I refactored as mentioned in my last comment above this should also be fixed. Rather, there wouldn't even be a need to compute numLabels and matchNil for the disjunction case if we hold a disjunction of selectorWrappers instead of a disjunction of label.Selector